### PR TITLE
Aparna hcl dev

### PR DESCRIPTION
**Pull request by aparnasasidharan2017**
I have added three data structures to HCL. Do you want to merge them with develop?

### CODE REVIEW

**Comment by hariharan-devarajan** [comment]
@aparnasasidharan2017 does these work with dynamic values as containers such as https://github.com/hariharan-devarajan/hcl/blob/dev/test/unordered_map_string_test.cpp#L121 I use dynamic strings?

**Comment by hariharan-devarajan** [comment]
@aparnasasidharan2017 Also squash all the commits into one. https://www.internalpointers.com/post/squash-commits-into-one-git

**Review by hariharan-devarajan — CHANGES_REQUESTED** [review]
I have left some comments to improve the tests and squashing all commits into 1.

**Comment by aparnasasidharan2017** [comment]
@hariharan-devarajan I have modified the testcases to cover shared-memory throughput and rpc throughput. My implementations do not use boost interprocess memory mapping for on-node rpc calls. I have assumed the implementation would use libfabric + shm instead of sockets for on-node communication. So, there are two code sections in the test cases : local function calls on servers and rpc calls from clients to servers.

**Comment by hariharan-devarajan** [comment]
> @hariharan-devarajan I have modified the testcases to cover shared-memory throughput and rpc throughput. My implementations do not use boost interprocess memory mapping for on-node rpc calls. I have assumed the implementation would use libfabric + shm instead of sockets for on-node communication. So, there are two code sections in the test cases : local function calls on servers and rpc calls from clients to servers.

the boost interprocess is only used to do node-local shared memory calls on the shared memory created by the servers. So in my tests, I have the following three blocks.

1. a local data structure (std implementations) as baseline to understand performance of your distributed data structure.
2. performance of your data structure on shared memory created by the server on that node. For this, you can hash the requests to the node-local servers so that they do not do RPC calls but do update directly on shared memory.
3. performance of your data structure across nodes, which uses rpc to communicate to the server and then uses shm calls to update the shared memory region on that node.

Whether you use direct shm or boost interprocess, all three cases would apply.

**Comment by hariharan-devarajan** [comment]
@aparnasasidharan2017 Please address the API concern I have Then i will merge the PR

**Comment by aparnasasidharan2017** [comment]
@hariharan-devarajan My testcases have 2 & 3 (shm and RPC). I didn't add a  baseline stl implementation because it will have to be single threaded or under a coarse lock. I'll add stl+coarse locks as baseline.

**Review by hariharan-devarajan — CHANGES_REQUESTED** [review]

File: test/skiplist_test.cpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    if(client_rank < rem) num_ops++;
+
+    MPI_Barrier(client_comm);
+
+    auto t1 = std::chrono::high_resolution_clock::now();
+    for (int i = 0; i < num_ops; i++) 
+    {
+	int op = random()%3;
+        int value = dist(rd)%10000;
+        uint64_t k = s->serverLocation(value);
```
[the remark this answers is not in the document]
Comment (aparnasasidharan2017) [reply test/skiplist_test.cpp]: I have made the changes. It makes sense for hashtables and skiplist.

**Comment by hariharan-devarajan** [comment]
@aparnasasidharan2017 Thank you for your patience with the PR. It looks good now.

**Pull request merged**